sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/lang/SqlDelightFoldingBuilder.kt

```diff
@@ -0,0 +1,98 @@
+import com.squareup.sqldelight.core.psi.SqlDelightStmtIdentifier
+import com.squareup.sqldelight.intellij.util.childOfType
+import com.squareup.sqldelight.intellij.util.childrenWithType
+import com.squareup.sqldelight.intellij.util.nextLeafOrNull
+import com.squareup.sqldelight.intellij.util.prevSiblingOfType
+
+class SqlDelightFoldingBuilder : FoldingBuilder, DumbAware {
+
+  override fun buildFoldRegions(root: ASTNode, document: Document): Array<FoldingDescriptor> {
+    return root.createFoldingDescriptors()
+  }
+
+  private fun ASTNode.createFoldingDescriptors(): Array<FoldingDescriptor> {
+    val descriptors = mutableListOf<FoldingDescriptor>()
+    val importElements = mutableListOf<PsiElement>()
+    val statementLists = childrenWithType(SqlDelightTypes.SQL_STMT_LIST)
+    for (statementList in statementLists) {
+      for (statement in statementList.getChildren(null)) {
+        when (statement.elementType) {
+          SqlDelightTypes.IMPORT_STMT -> importElements += statement.psi
+          SqliteTypes.STATEMENT -> {
+            val psi = statement.psi
+            val sqlStatement = statement.firstChildNode?.firstChildNode
+            val descriptor = if (sqlStatement?.elementType == SqliteTypes.CREATE_TABLE_STMT) {
+              psi.toCreateTableDescriptor(sqlStatement?.psi)
+            } else {
+              psi.toStatementDescriptor(psi.prevSiblingOfType<SqlDelightStmtIdentifier>())
+            }
+            if (descriptor != null) {
+              descriptors += descriptor
+            }
+          }
+        }
+      }
+    }
+    descriptors += importElements.toImportDescriptors()
+    return descriptors.toTypedArray()
+  }
+
+  private fun PsiElement.toCreateTableDescriptor(createTableStmt: PsiElement?): FoldingDescriptor? {
+    val openingBraceElement =
+        createTableStmt?.firstChild?.nextLeafOrNull { text == "(" } ?: return null
+    val start = openingBraceElement.textRange.startOffset
+    val end = lastChild.textRange.endOffset
+    return NamedFoldingDescriptor(this, start, end, null, "(...);")
+  }
+
+  private fun PsiElement.toStatementDescriptor(stmtIdentifier: PsiElement?): FoldingDescriptor? {
+    if (stmtIdentifier == null) return null
+    val start = stmtIdentifier.textRange.endOffset
+    val end = lastChild.textRange.endOffset
+    return NamedFoldingDescriptor(this, start, end, null, "...")
+  }
+
+  private fun List<PsiElement>.toImportDescriptors(): List<FoldingDescriptor> {
+
+    fun toDescriptor(): FoldingDescriptor? {
+      val whitespaceElement = first().childOfType<PsiWhiteSpace>() ?: return null
+      val start = whitespaceElement.textRange.endOffset
+      val end = last().textRange.endOffset
+      return NamedFoldingDescriptor(first(), start, end, null, "...")
+    }
+
+    return if (size > 1) listOfNotNull(toDescriptor()) else emptyList()
+  }
+
+  override fun getPlaceholderText(node: ASTNode) = "..."
+
+  override fun isCollapsedByDefault(node: ASTNode) = false
```

Comment:
Ideally this should be true for import lists, but we currently don't have a rule in the grammar for import lists, worth introducing?